File: implementation/scipamato/core/web/src/test/java/ch/difty/scipamato/core/web/authentication/LoginPageTest.java
package ch.difty.scipamato.core.web.authentication;

import com.giffing.wicket.spring.boot.starter.configuration.extensions.external.spring.security.SecureWebSession;
import de.agilecoders.wicket.core.markup.html.bootstrap.button.BootstrapButton;
import org.apache.wicket.markup.html.form.Form;
import org.apache.wicket.markup.html.internal.HtmlHeaderContainer;
import org.apache.wicket.request.mapper.parameter.PageParameters;
import org.apache.wicket.util.tester.FormTester;
import org.junit.jupiter.api.Test;

import ch.difty.scipamato.core.web.WicketTest;

@SuppressWarnings("SpellCheckingInspection")
class LoginPageTest extends WicketTest {

    @Test
    void assertPage() {
        getTester().startPage(makePage());
        getTester().assertRenderedPage(getPageClass());

        assertSpecificComponents();

        getTester().assertNoErrorMessage();
        getTester().assertNoInfoMessage();
    }

    private LoginPage makePage() {
        return new LoginPage(new PageParameters());
    }

    private Class<LoginPage> getPageClass() {
        return LoginPage.class;
    }

    private void assertSpecificComponents() {
        String b = "form";
        getTester().assertComponent(b, Form.class);
        assertLabeledTextField(b, "username");
        assertLabeledTextField(b, "password");
        getTester().assertComponent(b + ":signin", BootstrapButton.class);

        assertMenuEntries();
    }

    @Test
    void whenSigningIn_withWrongCredentials_providesErrorMessage() {
        SecureWebSession session = (SecureWebSession) getTester().getSession();
        session.signOut();
        getTester().startPage(LoginPage.class);
        FormTester formTester = getTester().newFormTester("form");
        formTester.setValue("username", "testuser");
        formTester.setValue("password", "wrongpw");
        formTester.submit();
        getTester().assertErrorMessages("The username and/or password were invalid.");
        getTester().assertRenderedPage(getPageClass());
    }

    private void assertMenuEntries() {
        getTester().assertComponent("_header_", HtmlHeaderContainer.class);
        getTester().assertInvisible("navbar");
    }
}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/web/authentication/LoginPageTest.kt
package ch.difty.scipamato.publ.web.authentication

import ch.difty.scipamato.publ.web.WicketTest
import com.giffing.wicket.spring.boot.starter.configuration.extensions.external.spring.security.SecureWebSession
import de.agilecoders.wicket.core.markup.html.bootstrap.button.BootstrapButton
import org.apache.wicket.markup.html.form.Form
import org.apache.wicket.markup.html.internal.HtmlHeaderContainer
import org.apache.wicket.request.mapper.parameter.PageParameters
import org.junit.jupiter.api.Test

class LoginPageTest : WicketTest() {

    @Test
    fun assertPage() {
        tester.startPage(makePage())
        tester.assertRenderedPage(pageClass)
        assertSpecificComponents()
        tester.assertNoErrorMessage()
        tester.assertNoInfoMessage()
    }

    private fun makePage(): LoginPage = LoginPage(PageParameters())

    private val pageClass: Class<LoginPage>
        get() = LoginPage::class.java

    private fun assertSpecificComponents() {
        val b = "form"
        tester.assertComponent(b, Form::class.java)
        assertLabeledTextField(b, "username")
        assertLabeledTextField(b, "password")
        tester.assertComponent("$b:signin", BootstrapButton::class.java)
        assertMenuEntries()
    }

    @Test
    fun whenSigningIn_withWrongCredentials_providesErrorMessage() {
        val session = tester.session as SecureWebSession
        session.signOut()
        tester.startPage(LoginPage::class.java)
        val formTester = tester.newFormTester("form")
        formTester.setValue("username", "testuser")
        formTester.setValue("password", "wrongpw")
        formTester.submit()
        tester.assertErrorMessages("The username and/or password were invalid.")
        tester.assertRenderedPage(pageClass)
    }

    private fun assertMenuEntries() {
        tester.assertComponent("_header_", HtmlHeaderContainer::class.java)
        tester.assertInvisible("navbar")
    }
}
